cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerSyncApiCommandIdProvider.java

```diff
@@ -0,0 +1,136 @@
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+import com.cloudera.api.swagger.ClustersResourceApi;
+import com.cloudera.api.swagger.client.ApiException;
+import com.cloudera.api.swagger.model.ApiCommand;
+import com.sequenceiq.cloudbreak.cm.commands.SyncApiCommandPollerConfig;
+import com.sequenceiq.cloudbreak.cm.commands.SyncApiCommandRetriever;
+import com.sequenceiq.cloudbreak.cm.polling.ClouderaManagerPollingServiceProvider;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.logger.MDCBuilder;
+import com.sequenceiq.cloudbreak.service.CloudbreakException;
+
+@Service
+public class ClouderaManagerSyncApiCommandIdProvider {
+
+    private static final Logger LOGGER =
+            LoggerFactory.getLogger(ClouderaManagerSyncApiCommandIdProvider.class);
+
+    private final SyncApiCommandPollerConfig syncApiCommandPollerConfig;
+
+    private final SyncApiCommandRetriever syncApiCommandRetriever;
+
+    private final ClouderaManagerPollingServiceProvider clouderaManagerPollingServiceProvider;
+
+    private final ExecutorService executorService;
+
+    public ClouderaManagerSyncApiCommandIdProvider(
+            SyncApiCommandRetriever syncApiCommandRetriever,
+            ClouderaManagerPollingServiceProvider clouderaManagerPollingServiceProvider,
+            SyncApiCommandPollerConfig syncApiCommandPollerConfig,
+            @Qualifier("cloudbreakListeningScheduledExecutorService")
+                    ExecutorService executorService) {
+        this.syncApiCommandRetriever = syncApiCommandRetriever;
+        this.executorService = executorService;
+        this.clouderaManagerPollingServiceProvider = clouderaManagerPollingServiceProvider;
+        this.syncApiCommandPollerConfig = syncApiCommandPollerConfig;
+    }
+
+    /**
+     * Obtain a sync API client config command ID with the following steps:
+     * 1. check are there any currently running commands -> stop if there is
+     * 2. check are there any existing successful sync API commands (that have finished already) - to collect wrong command IDs
+     * 3. execute the sync API command with a timeout
+     * 4. if the API command ties out, use 1. with polling
+     * 5. if polling finished, do a last check, returns with the command id or throw an exception
+     */
+    public BigDecimal executeSyncApiCommandAndGetCommandId(String commandName, ClustersResourceApi api, Stack stack,
+            List<ApiCommand> activeCommands, Callable<ApiCommand> callable) throws CloudbreakException, ApiException {
```

Comment:
`callable` is not a meaningful parameter name. please rename it here and in the other methods affected too